src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -295,11 +295,17 @@ private void ConsumeFromRemainingBuffer(int bytesToConsume)
             {
                 Debug.Assert(_pool.UsingProxy);
 
+                // TODO: #28863 Part 1: Uri.IdnHost should include [] around IPv6 address.
                 // If the hostname is an IPv6 address, uri.IdnHost will return the address without enclosing [].
                 // In this case, use uri.Host instead, which will correctly enclose with [].
                 // Note we don't need punycode encoding if it's an IP address, so using uri.Host is fine.
-                await WriteAsciiStringAsync(uri.HostNameType == UriHostNameType.IPv6 ?
-                    uri.Host : uri.IdnHost).ConfigureAwait(false);
+
+                // TODO: #28863 Part 2: Uri.Host LLA (Link-local address) for IPv6 address doesn't contain %number part.
+                // Since scope is mandatory for LLA, we will use uri.IdnHost for IPv6 LLA address.
+                // From RFC 4291, IPv6 LLA has formate like: fe80::/10.
+                await WriteAsciiStringAsync((uri.HostNameType == UriHostNameType.IPv6 &&
+                    uri.IdnHost.IndexOf("fe80") != 0) ?
+                        uri.Host : uri.IdnHost).ConfigureAwait(false);
```

Comment:
Why not just do:
```C#
if (uri.HostNameType == UriHostNameType.IPv6)
{
    await WriteByteAsync((byte)'[').ConfigureAwait(false);
    await WriteAsciiStringAsync(uri.IdnHost).ConfigureAwait(false);
    await WriteByteAsync((byte)']').ConfigureAwait(false);
}
else
{
    await WriteAsciiStringAsync(uri.IdnHost).ConfigureAwait(false);
}
```
?

Reply:
> Why are we limiting it to fe80?

I copied the logic from [`IPAddress.IsIPv6LinkLocal`](https://github.com/dotnet/corefx/blob/008c119440aa5219cc3c274a3a1edfa8c21711f3/src/System.Net.Primitives/src/System/Net/IPAddress.cs#L467) implementation. I can make it more strict and add tests to it.

> Why not just do:

This is definitely a better way. I didn't do that first because I want to make the change consistent, like below `isNonLLAIPv6Address` logic.

Reply:
According to the RFC for IPv6, "interface id" can be present for non LLA IPv6 addresses.

https://tools.ietf.org/html/rfc4291

It is required to be present for LLA because fe80:: addresses can be reused across multiple interfaces on a machine. So, the interface-id (a.k.a scope-id) is used to differentiate them.

Most common uses of IPv6 don't include the interface-id, though, for non-LLA addresses.